s2a/src/test/java/io/grpc/s2a/internal/handshaker/FakeS2AServerTest.java

```diff
@@ -79,6 +77,7 @@ public void callS2AServerOnce_getTlsConfiguration_returnsValidResult()
 
     try {
       S2AServiceGrpc.S2AServiceStub asyncStub = S2AServiceGrpc.newStub(channel);
+      CountDownLatch finishLatch = new CountDownLatch(1);
```

Comment:
I suggest using a ListenableFuture and completing it with the `response` within onCompleted() (or completing it with an error in onError()). That then guarantees you don't look at the response variable before it is set. (`response` can be a field within the StreamObserver).

Does that sound better? I agree that this finishLatch would work.

Reply:
I'm a bit confused on how we can do this so that we don't look at the `response` before it is set. What would we submit to the `ListenableFuture<SessionResp>`? I would think it's a function busy waiting until response is non-null. 

In https://github.com/grpc/grpc-java/pull/11673/commits/cd3a46271277cf236a54e29f8a9167beaa00cbef I think I implemented what you are suggesting (I think something is incorrect about the impl though because flakiness came back), but I may be misunderstanding something. Let me know if I am misinterpreting something.

Reply:
```java
      SettableFuture<SessionResp> respFuture = SettableFuture.create();
      StreamObserver<SessionReq> requestObserver =
          asyncStub.setUpSession(
              new StreamObserver<SessionResp>() {
                private SessionResp recvResp;

                @Override
                public void onNext(SessionResp resp) {
                  recvResp = resp;
                }

                @Override
                public void onError(Throwable t) {
                  respFuture.setException(t);
                }

                @Override
                public void onCompleted() {
                  respFuture.set(recvResp);
                }
              });
```

Reply:
Thanks for clarifying SettableFuture!
https://github.com/grpc/grpc-java/pull/11673/commits/6945599600417fd0fdbda992c5d1c3fcf4e34803